npm/debugTools/dataplane/parser.go

```diff
@@ -0,0 +1,338 @@
+		curReadIndex = nextReadIndex
+		if len(line) == 0 {
+			continue
+		}
+		if bytes.HasPrefix(line, util.CommitBytes) || line[0] == '*' {
+			break
+		}
+		if line[0] == ':' && len(line) > 1 {
+			// We assume that the <line> contains space - chain lines have 3 fields,
+			// space delimited. If there is no space, this line will panic.
+			spaceIndex := bytes.Index(line, util.SpaceBytes)
+			if spaceIndex == -1 {
+				panic(fmt.Sprintf("Unexpected chain line in iptables-save output: %v", string(line)))
+			}
+			chainName := string(line[1:spaceIndex])
+			if val, ok := chainMap[chainName]; ok {
+				val.Data = line
+				chainMap[chainName] = val
+			} else {
+				chainMap[chainName] = &IptablesChain{Name: chainName, Data: line, Rules: make([]*IptablesRule, 0)}
+			}
+		} else if line[0] == '-' && len(line) > 1 {
+			// rules
+			chainName, ruleStartIndex := p.parseChainNameFromRuleLine(line)
+			val, ok := chainMap[chainName]
+			if !ok {
+				val = &IptablesChain{chainName, []byte{}, make([]*IptablesRule, 0)}
+			}
+			val.Rules = append(val.Rules, p.parseRuleFromLine(line[ruleStartIndex:]))
+		}
+	}
+	return chainMap
+}
+
+// parseLine parse the line starting from the given readIndex of the iptableBuffer.
+// Returns a slice of line starting from given read index and the next index to read from.
+func (p *Parser) parseLine(readIndex int, iptableBuffer []byte) ([]byte, int) {
+	curReadIndex := readIndex // index of iptableBuffer to start reading from
+
+	// consume left spaces
+	for curReadIndex < len(iptableBuffer) {
+		if iptableBuffer[curReadIndex] != ' ' {
+			break
+		}
+		curReadIndex++
+	}
+	leftLineIndex := curReadIndex           // start index of line
+	rightLineIndex := -1                    // end index of line
+	lastNonWhiteSpaceIndex := leftLineIndex // index of last seen non-white space character
+
+	for ; curReadIndex < len(iptableBuffer); curReadIndex++ {
+		switch iptableBuffer[curReadIndex] {
+		case ' ':
+			if rightLineIndex == -1 {
+				rightLineIndex = curReadIndex // update end index of line
+			}
+		case '\n':
+			if rightLineIndex == -1 { // if end index of line is not set
+				rightLineIndex = curReadIndex
+				if curReadIndex == len(iptableBuffer)-1 {
+					// if this is also the end of the buffer
+					return iptableBuffer[leftLineIndex:], curReadIndex + 1
+				}
+			}
+			// return line slice and the next index to read from
+			return iptableBuffer[leftLineIndex:rightLineIndex], curReadIndex + 1
+		default:
+			lastNonWhiteSpaceIndex = curReadIndex // update index of last non-white space character
+			rightLineIndex = -1                   // reset right index of line
+		}
+	}
+	// line with right spaces (unlikely to encounter)
+	return iptableBuffer[leftLineIndex : lastNonWhiteSpaceIndex+1], curReadIndex
+}
+
+// parseChainNameFromRuleLine gets the chain name from given rule line
+func (p *Parser) parseChainNameFromRuleLine(line []byte) (string, int) {
+	spaceIndex1 := bytes.Index(line, util.SpaceBytes)
```

Comment:
It seems ok to use only spaceIndex instead of using both spaceIndex1 and spaceIndex2.
Also it is ok to use only start and end. 
Slightly better to use chainStart and chainEnd?